phoenix-core/src/main/java/org/apache/phoenix/coprocessor/TaskRegionObserver.java

```diff
@@ -157,18 +191,58 @@ public static void addTask(PhoenixConnection conn, TaskType taskType, String ten
                     PhoenixDatabaseMetaData.TENANT_ID + ", " +
                     PhoenixDatabaseMetaData.TABLE_SCHEM + ", " +
                     PhoenixDatabaseMetaData.TABLE_NAME + " ) VALUES(?,?,?,?)");
-            stmt.setByte(1, taskType.getSerializedValue());
-            if (tenantId != null) {
-                stmt.setString(2, tenantId);
+            stmt = setValuesToAddTaskPS(stmt, taskType, tenantId, schemaName, tableName);
+        } catch (SQLException e) {
+            throw new IOException(e);
+        }
+        mutateSystemTaskTable(conn, stmt, accessCheckEnabled);
+    }
+
+    public static void addTask(PhoenixConnection conn, TaskType taskType, String tenantId, String schemaName,
```

Comment:
Seems like there should be a separation of concerns here -- persisting a Task to a Phoenix table and reading it back is distinct from scheduling a task. I'd suggest breaking some of these Task classes into their own files and either fully encapsulate their persistence logic in the Tasks (the Data Access Object pattern) or have standalone persistence classes that handle the persistence (the Repository pattern). 

Either way would improve the readability and reduce the coupling between the scheduling logic and the persistence logic. 

Also agree with @kadirozde 's ideas to use reflection and JSON parsing to reduce the amount of boilerplate necessary to hook in each new task type for this framework.

Reply:
@gjacoby126 This is not a new method, it was already in this file. I moved it to this class. I am not seeing the patterns for other System tables. I will add a Tasks class tho.